pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -0,0 +1,196 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.io.util;
+
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import org.apache.pinot.common.utils.StringUtil;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+/**
+ * Implementation of {@link ValueReader} that will allow each value to be of variable
+ * length and there by avoiding the unnecessary padding.
+ *
+ * The layout of the file is as follows:
+ * <p> Header Section: </p>
+ * <ul>
+ *   <li> Magic header byte sequence. </li>
+ * </ul>
+ *
+ * <p> Values: </p>
+ * <ul>
+ *   <li> Integer offsets to start position of byte arrays. </li>
+ *   <li> All byte arrays. </li>
+ * </ul>
+ *
+ * Only sequential writes are supported.
+ *
+ * @see FixedByteValueReaderWriter
+ */
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
+
+  /**
+   * Header used to identify the dictionary files written in variable length bytes
+   * format. Keeping this a mix of alphanumeric with special character to avoid
+   * collisions with the regular int/string dictionary values written in fixed size
+   * format.
+   */
+  private static final byte[] MAGIC_HEADER = StringUtil.encodeUtf8("....vl1;");
```

Comment:
we need to think of a better way to store the magic header. See Gzip for  inspiration. https://en.wikipedia.org/wiki/Gzip 
```a 10-byte header, containing a magic number (1f 8b), compression ID (08 for DEFLATE), file flags, a 32-bit timestamp, compression flags and operating system ID.```

Reply:
I would include a version number either in the header or following the header. This will allow us to auto-detect dictionary types on load, and evolve it

Reply:
I've added version number in the header now which should let us evolve the store.